### PR TITLE
Fix x-discriminator-value processing

### LINKED ISSUE #17343: [BUG] [JAVA] Generated @JsonTypeName conflicts with value of x-discriminator-value
#### Bug Report Checklist

- [ ] Have you provided a full/minimal spec to reproduce the issue?
- [ ] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [ ] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [ ] Have you searched for related issues/PRs?
- [ ] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

[BUG] [JAVA] Generated @JsonTypeName doesn't use the value of x-discriminator-value

##### Description

@JSonTypeName uses {{name}} instead of x-discriminator-value if it exists.

Expected:
no @JSonTypeName annotation
if you remove @JsonTypeName from template everything works fine (Jackson seems to retrieve typeName from parent).

public class DogItemApiDTO extends PetItemApiDTO {

Actual:
@JsonTypeName("DogItem")
public class DogItemApiDTO extends PetItemApiDTO {

##### openapi-generator version 7.1.0
regression since 6.4.0.  Maybe fix for #14731

##### OpenAPI declaration file content or url

```yaml
swagger: "2.0"
info:
  version: 1.0.0
  title: x-discriminator-value bug
paths:
  /getItems:
    get:
      description: test
      operationId: getAllPets
      responses:
        200:
          description: 'list of PetItems'
          schema:
            type: array
            items:
              $ref: '#/definitions/PetItem'
definitions:
  PetType:
    type: string
    enum:
      - DOG
      - CAT
  PetItem:
    discriminator: type
    properties:
      identifier:
        type: string
      type:
        $ref: '#/definitions/PetType'

  DogItem:
    x-discriminator-value: DOG
    allOf:
      - $ref: '#/definitions/PetItem'
      - type: object
        properties:
          packSize:
            type: integer
  CatItem:
    x-discriminator-value: CAT
    allOf:
      - $ref: '#/definitions/PetItem'
      - type: object
        properties:
          huntingSkill:
            type: string

```

##### Generation Details

openapi-generator-cli generate -i swaggerDiscriminator.yaml -g spring --model-name-suffix=ApiDto

##### Steps to reproduce

Contract in swagger-2
inheritance with a discriminator
use of x-discriminator-value
modelNameSuffix

Use the spring generator.

##### Related issues/PRs

#10822

##### Suggest a fix

Update pojo.mustache to use the vendorExtensions.x-discriminator-value if it exists

{{^hasDiscriminatorWithNonEmptyMapping}}
@JsonTypeName("{{^vendorExtensions.x-discriminator-value}}{{name}}{{/vendorExtensions.x-discriminator-value}}{{#vendorExtensions.x-discriminator-value}}{{vendorExtensions.x-discriminator-value}}{{/vendorExtensions.x-discriminator-value}}")
{{/hasDiscriminatorWithNonEmptyMapping}}

### PR DESCRIPTION
**Pull request by jpfinne**
Fix #17343 allowing correct marshalling/unmarshalling of objects using an x-discrimininator-value

Test various combination of swagger/openapi, modelNameSuffix...

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [7.3.0]
- [x] If your PR is targeting a particular programming language, @mention the [technical committee]@cachescrubber (2022/02) @welshm (2022/02) @MelleD (2022/02) @atextor (2022/02) @manedev79 (2022/02) @javisst (2022/02) @borsch (2022/02) @banlevente (2022/02) @Zomzog (2022/09) @martin-mfg (2023/08)
, so they are more likely to review the pull request.

### CODE REVIEW

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java [hunk 1]
```diff
@@ -419,7 +420,8 @@ public CodegenDiscriminator getDiscriminator() {
 
     public void setDiscriminator(CodegenDiscriminator discriminator) {
         this.discriminator = discriminator;
-        if (discriminator != null && !discriminator.getMappedModels().isEmpty()) {
+        if (discriminator != null && (!discriminator.getMappedModels().isEmpty()) || this.vendorExtensions.containsKey("x-discriminator-value")) {
```
Comment (martin-mfg) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java:423]: Thanks for the PR!

The 2nd condition on this line is surrounded by unnecessary parenthesis. So I want to check - did you mean to write `if (discriminator != null && (!discriminator.getMappedModels().isEmpty() || this.vendorExtensions.containsKey("x-discriminator-value"))) {` (with parenthesis around the 2nd and 3rd condition) instead?
If it's not a mistake and you simply added the parenthesis for clarity, that's fine for me.

Comment (jpfinne) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java:423]: @martin-mfg  
Thanks for the review.

**Clarity**.  only processing if there is a discriminator.

In the context of the method:
if x-discriminator-value is defined, discriminator is not null.  So indeed, the parenthesis are not strictly necessary.

**Comment by wing328** [comment]
cc @OpenAPITools/generator-core-team

**Comment by wing328** [comment]
can you please update the samples too? https://github.com/OpenAPITools/openapi-generator/actions/runs/7940713183/job/21756786029?pr=17781

**Comment by jpfinne** [comment]
@wing328  Sorry I'm new to the whole process.
I've managed to install a portable version of git bash on my Windows laptop.  And I can regenerate the samples.

[JacksonTest on master](https://github.com/OpenAPITools/openapi-generator/blob/master/samples/server/petstore/springboot/src/test/java/org/openapitools/JacksonTest.java) is wrong in the test for
```
returns("CatDto", AnimalDto::getClassName)
```
It should be "Cat".

All the previous attempts tried to skip the generation of  ```@JsonTypeName```.  It does not work for all cases (parent class, middle class...)

I've used here a new approach:  the actual value in @JsonTypeName is set in a postProcessor


One native test does fail. I don't know why.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java [hunk 2]
```diff
@@ -345,6 +345,15 @@ public boolean getIsClassnameSanitized() {
...
 
+    /**
+     *
+     * @return true if @JSonTypeName is needed
+     */
+    public boolean getUseJsonTypeName() {
+        Object xDiscriminatorValue = vendorExtensions.get("x-discriminator-value");
+        return getIsClassnameSanitized() || (xDiscriminatorValue!=null && !classname.equals(xDiscriminatorValue));
+    }
+
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java:356]: This code block is for Java only so I don't think we should include it in Codegen Model.

What about moving it to Abstract Java Codegen instead?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java [hunk 3]
```diff
@@ -668,10 +668,38 @@ public void processOpts() {
         additionalProperties.put(CONTAINER_DEFAULT_TO_NULL, containerDefaultToNull);
     }
 
+    private void postProcessMappingDiscriminators(Map<String, ModelsMap> objs) {
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java:671]: can you please add a docstring explaining what this function does?

add a bit more code comment below explaining what the for loops do would be helpful.

**Comment by jpfinne** [comment]
@wing328  thank you for your review.

The latest commit uses a lambda for useJsonTypeName.

I'm a bit scared by the amount of samples updated.  
There are updated in rust samples.  I don't know why.

btw I use a portable version of Gitbash on Windows.  It generates cr/lf for FILES.  Is there a way to configure the EOL?

**Review by martin-mfg — CHANGES_REQUESTED** [review]

File: modules/openapi-generator/src/test/java/org/openapitools/codegen/java/spring/SpringCodegenTest.java [hunk 4]
```diff
@@ -4396,6 +4366,103 @@ public void givenMultipartForm_whenGenerateBlockedServer_thenParameterAreCreated
...
+
+    private void expectJsonTypeNameAnnotation(File file, String expectedExtends, String expectedJsonAnnotation) {
+        try {
+        JavaFileAssert javaFileAssert = JavaFileAssert.assertThat(file);
+        if (expectedExtends != null) {
+            javaFileAssert.fileContains(" extends " + expectedExtends);
+        } else {
+            javaFileAssert.doesNotHaveToString(" extends ");
+        }
+        TypeAnnotationAssert dogAnnotation = javaFileAssert.assertTypeAnnotations();
```
[the remark this answers is not in the document]
Comment (jpfinne) [reply modules/openapi-generator/src/test/java/org/openapitools/codegen/java/spring/SpringCodegenTest.java:4454]: You're right.  I've fixed it

File: modules/openapi-generator/src/test/java/org/openapitools/codegen/java/spring/SpringCodegenTest.java [hunk 5]
```diff
@@ -1661,45 +1670,6 @@ public void testOneOfAndAllOf() throws IOException {
         assertFileContains(Paths.get(outputPath + "/src/main/java/org/openapitools/model/PizzaSpeziale.java"), "import java.math.BigDecimal");
     }
 
-    @Test
```
[the remark this answers is not in the document]
Comment (jpfinne) [reply modules/openapi-generator/src/test/java/org/openapitools/codegen/java/spring/SpringCodegenTest.java:1664]: @martin-mfg   The generated PizzaSpeziale.java contains the import.  It should not.  
But a fix for this minor bug is outside the scope of this PR.

More important:  it does not override  getPizzaSize()

[the remark this answers is not in the document]
Comment (martin-mfg) [reply modules/openapi-generator/src/test/java/org/openapitools/codegen/java/spring/SpringCodegenTest.java:1664]: I guess you are referring to the test "testOneOfAndAllOf". My comment was meant to be about the test "testDiscriminatorWithMappingIssue14731". Could you restore and adapt that one?

[the remark this answers is not in the document]
Comment (jpfinne) [reply modules/openapi-generator/src/test/java/org/openapitools/codegen/java/spring/SpringCodegenTest.java:1664]: Ok, I see the issue.  It used he standard library (okhttp-gson). So the test was meaningless.
I've change the library to resttemplate and now `@jsonTypeName` are present.

[the remark this answers is not in the document]
Comment (martin-mfg) [reply modules/openapi-generator/src/test/java/org/openapitools/codegen/java/spring/SpringCodegenTest.java:1664]: You made the changes in `JavaClientCodegenTest.java`, but they still need to be done in `SpringCodegenTest.java`, correct?

[the remark this answers is not in the document]
Comment (jpfinne) [reply modules/openapi-generator/src/test/java/org/openapitools/codegen/java/spring/SpringCodegenTest.java:1664]: I've added additial assertions

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java [hunk 6]
```diff
@@ -2492,4 +2537,40 @@ public static void addImports(List<Map<String, String>> imports, CodegenModel cm
...
+     */
+    private static class UseJsonTypeName implements Mustache.Lambda {
+
+        /**
+         * Write the enclosed fragment if the json type name is different than the class name.
+         */
+        @Override
+        public void execute(Template.Fragment fragment, Writer writer) throws IOException {
+            // context index 1 is only valid when used in pojo.mustache
+            CodegenModel codegenModel = (CodegenModel)fragment.context(1);
```
Comment (martin-mfg) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java:2568]: Updating samples via `./bin/generate-samples.sh ./bin/configs/*.yaml` generates an error. The reason is that the cast in this line doesn't work in one of the sample cases.
> org.openapitools.codegen.model.ModelMap cannot be cast to class org.openapitools.codegen.CodegenModel

Comment (jpfinne) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java]: Sorry my mistake.  I used an obsolete version of the cli to generate the samples.
I' ve regenerated them with a fix for the cast.

**Comment by martin-mfg** [comment]
It would be great if @wing328 or someone from the Java technical committee could take over reviewing this PR; both to add another perspective and because I currently can't work on this.

**Comment by jpfinne** [comment]
> It would be great if @wing328 or someone from the Java technical committee could take over reviewing this PR; both to add another perspective and because I currently can't work on this.

@wing328  now what?

**Pull request closed without merging**